From ffd0b54edca4ae79a1da1489decd4cd0cf9f930a Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Bernhard=20=C3=9Cbelacker?= <bernhardu@mailbox.org>
Date: Mon, 21 Oct 2024 11:09:09 +0200
Subject: [PATCH 70/86] rpcrt4: Do not call strdup with a null pointer.

This gets visible when ASAN replaces the strdup implementation,
which does not handle the NULL case like the msvcrt implementation.
---
 dlls/rpcrt4/rpc_binding.c   | 10 +++++-----
 dlls/rpcrt4/rpc_transport.c |  4 ++--
 2 files changed, 7 insertions(+), 7 deletions(-)

diff --git a/dlls/rpcrt4/rpc_binding.c b/dlls/rpcrt4/rpc_binding.c
index d4e35bdf116..73f7d3f1d2e 100644
--- a/dlls/rpcrt4/rpc_binding.c
+++ b/dlls/rpcrt4/rpc_binding.c
@@ -979,11 +979,11 @@ RPC_STATUS RPC_ENTRY RpcBindingCopy(
 
   DestBinding->ObjectUuid = SrcBinding->ObjectUuid;
   DestBinding->BlockingFn = SrcBinding->BlockingFn;
-  DestBinding->Protseq = strdup(SrcBinding->Protseq);
-  DestBinding->NetworkAddr = strdup(SrcBinding->NetworkAddr);
-  DestBinding->Endpoint = strdup(SrcBinding->Endpoint);
-  DestBinding->NetworkOptions = wcsdup(SrcBinding->NetworkOptions);
-  DestBinding->CookieAuth = wcsdup(SrcBinding->CookieAuth);
+  DestBinding->Protseq = SrcBinding->Protseq ? strdup(SrcBinding->Protseq) : NULL;
+  DestBinding->NetworkAddr = SrcBinding->NetworkAddr ? strdup(SrcBinding->NetworkAddr) : NULL;
+  DestBinding->Endpoint = SrcBinding->Endpoint ? strdup(SrcBinding->Endpoint) : NULL;
+  DestBinding->NetworkOptions = SrcBinding->NetworkOptions ? wcsdup(SrcBinding->NetworkOptions) : NULL;
+  DestBinding->CookieAuth = SrcBinding->CookieAuth ? wcsdup(SrcBinding->CookieAuth) : NULL;
   if (SrcBinding->Assoc) SrcBinding->Assoc->refs++;
   DestBinding->Assoc = SrcBinding->Assoc;
 
diff --git a/dlls/rpcrt4/rpc_transport.c b/dlls/rpcrt4/rpc_transport.c
index 8d53e7cee35..968bcfa37b1 100644
--- a/dlls/rpcrt4/rpc_transport.c
+++ b/dlls/rpcrt4/rpc_transport.c
@@ -3302,8 +3302,8 @@ RPC_STATUS RPCRT4_CreateConnection(RpcConnection** Connection, BOOL server,
   NewConnection->ref = 1;
   NewConnection->server = server;
   NewConnection->ops = ops;
-  NewConnection->NetworkAddr = strdup(NetworkAddr);
-  NewConnection->Endpoint = strdup(Endpoint);
+  NewConnection->NetworkAddr = NetworkAddr ? strdup(NetworkAddr) : NULL;
+  NewConnection->Endpoint = Endpoint ? strdup(Endpoint) : NULL;
   NewConnection->NetworkOptions = wcsdup(NetworkOptions);
   NewConnection->CookieAuth = wcsdup(CookieAuth);
   NewConnection->MaxTransmissionSize = RPC_MAX_PACKET_SIZE;
-- 
2.47.1

